third_party/move/move-vm/runtime/src/storage/ty_layout_converter.rs

```diff
@@ -1,71 +1,126 @@
+#[derive(Debug)]
+pub struct LayoutWithDelayedFields {
+    layout: MoveTypeLayout,
+    contains_delayed_fields: bool,
+}
 
-        let mut count = 0;
-        self.type_to_type_layout_impl(ty, &mut count, 1)
-            .map(|(l, _)| l)
+impl LayoutWithDelayedFields {
+    /// If layout contains delayed fields, returns [None]. If there are no delayed fields, the
+    /// layout is returned.
+    pub fn into_layout_when_has_no_delayed_fields(self) -> Option<MoveTypeLayout> {
+        (!self.contains_delayed_fields).then_some(self.layout)
     }
 
-    /// Converts a runtime type to a type layout.
-    fn type_to_type_layout_with_identifier_mappings(
-        &self,
-        ty: &Type,
-    ) -> PartialVMResult<(MoveTypeLayout, bool)> {
-        let _timer = VM_TIMER.timer_with_label("Loader::type_to_type_layout");
+    /// If layout does not contain delayed fields, returns [None]. If there are delayed fields, the
+    /// layout is returned.
+    pub fn layout_when_contains_delayed_fields(&self) -> Option<&MoveTypeLayout> {
+        self.contains_delayed_fields.then_some(&self.layout)
+    }
 
-        let mut count = 0;
-        self.type_to_type_layout_impl(ty, &mut count, 1)
+    /// Unpacks and returns the layout and delayed fields flag for the caller to handle.
+    pub fn unpack(self) -> (MoveTypeLayout, bool) {
+        (self.layout, self.contains_delayed_fields)
     }
+}
 
-    /// Converts a runtime type to a fully annotated type layout, containing information about
-    /// field names.
-    fn type_to_fully_annotated_layout(&self, ty: &Type) -> PartialVMResult<MoveTypeLayout> {
-        let _timer = VM_TIMER.timer_with_label("Loader::type_to_type_layout");
+/// Converts runtime types to type layouts. The layout construction may load modules, and so, the
+/// functions may also charge gas.
+pub(crate) struct LayoutConverter<'a, T> {
+    struct_definition_loader: &'a T,
+}
 
-        let mut count = 0;
-        self.type_to_fully_annotated_layout_impl(ty, &mut count, 1)
+impl<'a, T> LayoutConverter<'a, T>
+where
+    T: StructDefinitionLoader,
+{
+    /// Creates a new layout converter with access to struct definition loader.
+    pub(crate) fn new(struct_definition_loader: &'a T) -> Self {
+        Self {
+            struct_definition_loader,
+        }
     }
-}
 
-// This is not intended to be implemented or used externally, so put abstract and other functions
-// into this crate trait.
-pub(crate) trait LayoutConverterBase {
-    fn vm_config(&self) -> &VMConfig;
-    fn fetch_struct_ty_by_idx(&self, idx: StructNameIndex) -> PartialVMResult<Arc<StructType>>;
-    fn struct_name_index_map(&self) -> &StructNameIndexMap;
+    /// Returns the layout of a type, as well as a flag if it contains delayed fields or not.
+    pub(crate) fn type_to_type_layout_with_delayed_fields(
+        &self,
+        gas_meter: &mut impl GasMeter,
+        traversal_context: &mut impl ModuleTraversalContext,
```

Comment:
I'm wondering instead of passing the traversal context all around, does it make more sense to store it together with GasMeter or something?

it's hard to reason about where the context comes from. also it's closely coupled with gas charging anyway